### PR TITLE
Support `allocated_ip_range` in `google_sql_database_instance`

### PR DESCRIPTION
**Pull request by shuyama1**
Fix https://github.com/hashicorp/terraform-provider-google/issues/9986
Also bring the test to GA as service networking is available in GA now

If this PR is for Terraform, I acknowledge that I have:

- [ ] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ ] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [ ] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [ ] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [ ] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
sql: added field `allocated_ip_range` to resource `google_sql_database_instance`
```

### CODE REVIEW

File: mmv1/third_party/terraform/website/docs/r/sql_database_instance.html.markdown [hunk 1]
```diff
@@ -295,6 +295,8 @@ This setting can be updated, but it cannot be removed after it is set.
 
 * `require_ssl` - (Optional) Whether SSL connections over IP are enforced or not.
 
+* `allocated_ip_range` - (Optional) The name of the allocated ip range for the private ip CloudSQL instance. For example: "google-managed-services-default". If set, the instance ip will be created in the allocated range. The range name must comply with [RFC 1035](https://datatracker.ietf.org/doc/html/rfc1035). Specifically, the name must be 1-63 characters long and match the regular expression [a-z]([-a-z0-9]*[a-z0-9])?. Reserved for future use.
```
Comment (rileykarson) [inline mmv1/third_party/terraform/website/docs/r/sql_database_instance.html.markdown:298]:
```suggestion
* `allocated_ip_range` - (Optional) The name of the allocated ip range for the private ip CloudSQL instance. For example: "google-managed-services-default". If set, the instance ip will be created in the allocated range. The range name must comply with [RFC 1035](https://datatracker.ietf.org/doc/html/rfc1035). Specifically, the name must be 1-63 characters long and match the regular expression [a-z]([-a-z0-9]*[a-z0-9])?.
```

I suspect we can drop this part.

File: mmv1/third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 2]
```diff
@@ -285,28 +278,34 @@ settings.backup_configuration.binary_log_enabled are both set to true.`,
...
 									"private_network": {
 										Type:             schema.TypeString,
 										Optional:         true,
 										ValidateFunc:     orEmpty(validateRegexp(privateNetworkLinkRegex)),
 										DiffSuppressFunc: compareSelfLinkRelativePaths,
-										AtLeastOneOf:     ipConfigurationKeys,
+										AtLeastOneOf:     []string{"settings.0.ip_configuration.0.authorized_networks", "settings.0.ip_configuration.0.ipv4_enabled", "settings.0.ip_configuration.0.require_ssl", "settings.0.ip_configuration.0.private_network"},
 										Description:      `The VPC network from which the Cloud SQL instance is accessible for private IP. For example, projects/myProject/global/networks/default. Specifying a network enables private IP. At least ipv4_enabled must be enabled or a private_network must be configured. This setting can be updated, but it cannot be removed after it is set.`,
 									},
+									"allocated_ip_range": {
```
Comment (rileykarson) [inline mmv1/third_party/terraform/resources/resource_sql_database_instance.go.erb:303]: If this isn't specified, is one automatically assigned?

**Review by rileykarson — COMMENTED** [review]
VCR's freaking out- are you able to start an independent run, or share a successful test log here?

File: mmv1/third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 3]
```diff
@@ -48,13 +48,6 @@ var (
 		"settings.0.backup_configuration.0.transaction_log_retention_days",
 	}
 
-	ipConfigurationKeys = []string{
```
[the remark this answers is not in the document]
Comment (shuyama1) [reply mmv1/third_party/terraform/resources/resource_sql_database_instance.go.erb:51]: My concern is that `allocated_ip_range` should only be set as long as `private_network` also got a correct value being set, so `allocated_ip_range` should not be in that `AtLeastOneOf`, and `ipConfigurationKeys` may be confusing if it only contains the other four fields but not `allocated_ip_range` (or maybe it is not confusing, it was just my concern). Also is there a way to indicate that a field should be set only if some other fields are set as well?

File: mmv1/third_party/terraform/website/docs/r/sql_database_instance.html.markdown (hunk 1 above)
Comment (shuyama1) [reply mmv1/third_party/terraform/website/docs/r/sql_database_instance.html.markdown:298]: Yeah. I agree and will do!

File: mmv1/third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 2 above)
Comment (shuyama1) [reply mmv1/third_party/terraform/resources/resource_sql_database_instance.go.erb:303]: If so, the field should be `optional `+ `computed`, right? I currently did not add the `computed` and also not specify a value in the test and it did not throw me any errors, so I assume it means the field is not optional computed?

**Comment by shuyama1** [comment]
I run the test independently and it passed (https://ci-oss.hashicorp.engineering/buildConfiguration/GoogleCloudBeta_ProviderGoogleCloudBetaMmUpstream/226542), but it seems never passed in the VCR test

File: mmv1/third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 3 above)
[the remark this answers is not in the document]
Comment (rileykarson) [reply mmv1/third_party/terraform/resources/resource_sql_database_instance.go.erb:51]: > My concern is that allocated_ip_range should only be set as long as private_network also got a correct value being set, so allocated_ip_range should not be in that AtLeastOneOf, and ipConfigurationKeys may be confusing if it only contains the other four fields but not allocated_ip_range (or maybe it is not confusing, it was just my concern).
Using AtLeastOneOf makes up for a specific deficiency in Terraform, I think it's fine to not represent the API 100% accurately here. The API will just return an error.

> Also is there a way to indicate that a field should be set only if some other fields are set as well?
Not that I can think of right now! It's _possible_ with `CustomizeDiff` but prone to errors and certainly not worth doing.

File: mmv1/third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 2 above)
Comment (rileykarson) [reply mmv1/third_party/terraform/resources/resource_sql_database_instance.go.erb:303]: Given this field is tied to using a private network, I think we'd need a test that uses a private network without setting `allocated_ip_range` to verify that. We don't have one right now (note, `TestAccSqlDatabaseInstance_settings_checkServiceNetworking` looks like a case where we do so, but the config is intentionally invalid so it won't successfully apply.)

File: mmv1/third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 3 above)
[the remark this answers is not in the document]
Comment (shuyama1) [reply mmv1/third_party/terraform/resources/resource_sql_database_instance.go.erb:51]: Yeah. I think the error returned is clear enough. Will bring back the `ipConfigurationKeys` and add `allocated_ip_range ` to it.

File: mmv1/third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 2 above)
Comment (shuyama1) [reply mmv1/third_party/terraform/resources/resource_sql_database_instance.go.erb:303]: Sure! I will add one test then.

**Review by rileykarson — APPROVED** [review]

**Pull request merged**